native/membrane_videocompositor_opengl_rust/src/scene.rs

```diff
@@ -211,31 +213,40 @@ impl Scene {
     ) -> Result<Self, CompositorError> {
         // FIXME: the shader should probably be constructed in this function instead of being passed as an argument
         Ok(Self {
-            videos: Vec::new(),
+            videos: BTreeMap::new(),
             render_target: YUVRenderTarget::new(out_width, out_height)?,
             shader_program,
         })
     }
 
     /// Add a video to the scene.
     /// The video will be placed in an area specified by `placement`
-    /// This returns the index that should be used for uploading textures to this video.
     pub fn add_video(
         &mut self,
+        video_idx: usize,
         placement: VideoPlacementTemplate,
         width: usize,
         height: usize,
-    ) -> Result<usize, CompositorError> {
-        self.videos.push(Video::new(placement, width, height)?);
-        Ok(self.videos.len() - 1)
+    ) -> Result<(), CompositorError> {
+        self.videos
+            .insert(video_idx, Video::new(placement, width, height)?);
+        Ok(())
+    }
+
+    /// Remove a video from the scene.
+    /// # Panics
+    ///
+    /// This function will panic if the `video_idx` is not an index of an existing video
+    pub fn remove_video(&mut self, video_idx: usize) {
+        self.videos.remove(&video_idx);
     }
 
     /// Upload a texture to a video specified by the `video_idx`
     ///
     /// # Panics
     ///
     /// This function will panic if the `video_idx` is not an index of an existing video and if `data` doesn't have proper length for a YUV420p-encoded frame for the specified video.
-    pub fn upload_texture(&self, video_idx: usize, data: &[u8]) -> Result<(), CompositorError> {
+    pub fn upload_texture(&self, video_idx: &usize, data: &[u8]) -> Result<(), CompositorError> {
         self.videos[video_idx].upload_texture(data)?;
         Ok(())
     }
```

Comment:
I really dislike the panics here, since a panic will (I believe) crash the BEAM entirely. I find it more acceptable in `draw_into`, since it only operates on objects created and validated internally in this NIF. Here we crash when a value passed down from elixir is incorrect, which I find more bad. Consider making a new `CompositorError` for passing an incorrect video index and returning it from here.

Also, we should later replace panics with errors everywhere later.

Reply:
also, passing a `usize` by reference instead of by value seems unnecessary.